src/batched/dense/impl/KokkosBatched_Gesv_Impl.hpp

```diff
@@ -0,0 +1,548 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+#ifndef __KOKKOSBATCHED_GESV_IMPL_HPP__
+#define __KOKKOSBATCHED_GESV_IMPL_HPP__
+
+/// \author Kim Liegeois (knliege@sandia.gov)
+
+#include "KokkosBatched_Util.hpp"
+#include <KokkosBatched_LU_Decl.hpp>
+#include "KokkosBatched_Trsm_Decl.hpp"
+
+namespace KokkosBatched {
+
+struct SerialStaticPivoting {
+  template <class MatrixType1, class MatrixType2, class VectorType1,
+            class VectorType2>
+  KOKKOS_INLINE_FUNCTION static void invoke(
+      const MatrixType1 A, const MatrixType2 PDAD, const VectorType1 Y,
+      const VectorType2 PDY, const VectorType2 D2, const VectorType2 tmp_v_1,
+      const VectorType2 tmp_v_2);
+};
+
+template <typename MemberType>
+struct TeamStaticPivoting {
+  template <class MatrixType1, class MatrixType2, class VectorType1,
+            class VectorType2>
+  KOKKOS_INLINE_FUNCTION static void invoke(
+      const MemberType &member, const MatrixType1 A, const MatrixType2 PDAD,
+      const VectorType1 Y, const VectorType2 PDY, const VectorType2 D2,
+      const VectorType2 tmp_v_1, const VectorType2 tmp_v_2);
+};
+
+template <typename MemberType>
+struct TeamVectorStaticPivoting {
+  template <class MatrixType1, class MatrixType2, class VectorType1,
+            class VectorType2>
+  KOKKOS_INLINE_FUNCTION static void invoke(
+      const MemberType &member, const MatrixType1 A, const MatrixType2 PDAD,
+      const VectorType1 Y, const VectorType2 PDY, const VectorType2 D2,
+      const VectorType2 tmp_v_1, const VectorType2 tmp_v_2);
+};
+
+template <class MatrixType1, class MatrixType2, class VectorType1,
+          class VectorType2>
+KOKKOS_INLINE_FUNCTION void SerialStaticPivoting::invoke(
+    const MatrixType1 A, const MatrixType2 PDAD, const VectorType1 Y,
+    const VectorType2 PDY, const VectorType2 D2, const VectorType2 tmp_v_1,
+    const VectorType2 tmp_v_2) {
+  using value_type = typename MatrixType1::non_const_value_type;
+  const int n      = A.extent(0);
+
+  for (int i = 0; i < n; ++i) {
+    D2(i)      = 0.;
+    tmp_v_1(i) = 0;
+    tmp_v_2(i) = 1.;
+    for (int j = 0; j < n; ++j) {
+      if (D2(i) < Kokkos::abs(A(j, i))) D2(i) = Kokkos::abs(A(j, i));
+      if (tmp_v_1(i) < Kokkos::abs(A(i, j))) tmp_v_1(i) = Kokkos::abs(A(i, j));
+    }
+    D2(i) = 1. / D2(i);
+  }
+
+  for (int i = 0; i < n; ++i) {
```

Comment:
Add a quick comment like `// scaling columns of A by their largest absolute value`
Also inverting these two loops might keep `D2(j)` in a register instead of streaming the values of `D2` from cache. This might already happen by compiler optimization?